source/robot_model/src/Model.cpp

```diff
@@ -29,17 +49,145 @@ bool Model::create_urdf_from_string(const std::string& urdf_string, const std::s
+      while ((start_position = urdf.find(target, start_position)) != std::string::npos) {
+          // Replace the target with the replacement string
+          urdf.replace(start_position, target.length(), replacement);
+          // Move past the last replacement
+          start_position += replacement.length();
+      }
+      package_paths.push_back(package_path);
+    }
+  }
+  return package_paths;
+}
+
 void Model::init_model() {
-  pinocchio::urdf::buildModel(this->get_urdf_path(), this->robot_model_);
+  std::ifstream file_stream(this->get_urdf_path());
+  if (!file_stream.is_open()) {
+      throw std::runtime_error("Unable to open file: " + this->get_urdf_path());
+  }
+  std::stringstream buffer;
+  buffer << file_stream.rdbuf();
+  auto urdf = buffer.str();
+
+  pinocchio::urdf::buildModelFromXML(urdf, this->robot_model_);
   this->robot_data_ = pinocchio::Data(this->robot_model_);
+
+  if (this->load_collision_geometries_) {
+    this->init_geom_model(urdf);
+  }
+
   // get the frames
   std::vector<std::string> frames;
   for (auto& f : this->robot_model_.frames) {
     frames.push_back(f.name);
   }
   // remove universe and root_joint frame added by Pinocchio
   this->frames_ = std::vector<std::string>(frames.begin() + 2, frames.end());
-  this->init_qp_solver();
+  this->init_qp_solver();  
+}
+
+// Method to initialize collision geometries
+void Model::init_geom_model(std::string urdf) {
+    try {
+      auto package_paths = this->resolve_package_paths_in_urdf(urdf);
+      pinocchio::urdf::buildGeom(this->robot_model_,
+                                std::istringstream(urdf),
+                                pinocchio::COLLISION, 
+                                this->geom_model_, 
+                                package_paths);
+      this->geom_model_.addAllCollisionPairs();
+    
+      std::vector<pinocchio::CollisionPair> excluded_pairs = this->generate_joint_exclusion_list();
+
+      // remove collision pairs for linked joints (i.e. parent-child joints)
+      for (const auto& pair : excluded_pairs) {
+          this->geom_model_.removeCollisionPair(pair);
+      }
+
+      this->geom_data_ = pinocchio::GeometryData(this->geom_model_);
+    } catch (const std::exception& ex) {
+        throw robot_model::exceptions::CollisionGeometryException("Failed to initialize geometry model for " + this->get_robot_name() + ": " + ex.what());
+    }
+}
+
+std::vector<pinocchio::CollisionPair> Model::generate_joint_exclusion_list() {
+    std::vector<pinocchio::CollisionPair> excluded_pairs;
+    // Iterate through all joints, except the universe joint (0), which has no parent
+    for (pinocchio::JointIndex joint_id = 1u; joint_id < static_cast<pinocchio::JointIndex>(this->robot_model_.njoints); ++joint_id){
+        // Find the parent joint of the current joint
+        pinocchio::JointIndex parent_id = this->robot_model_.parents[joint_id];
+        
+        // TODO: Replace this logic with actual geometry index lookup
```

Comment:
What is this TODO about? just for curiosity

Reply:
This approach assumes a one-to-one correspondence between joint indices and geometry indices, where each joint index directly corresponds to a geometry index with the same value. However, in more complex situations, the relationship between joints and their associated geometries in a robot model can be more complex for several reasons:

* **Multiple Geometries per Joint**
* **No Geometry for Some Joints**
* **Shared Geometries**
* **Non-Sequential Indexing between Geometries and Joint indicies**